### PR TITLE
Migrate to GitHub Actions; work around Dockerfile caching bug

### PR DESCRIPTION
**Pull request by glin**
There are two separate but related changes here:

1. Migrate from CircleCI to GitHub Actions. CircleCI has worked really well, but it seems like we're converging on GHA in the org, so less YAML config to learn means easier maintenance. The CircleCI config has also had a lot of copy-paste redundancy that made it a bit harder to maintain, and I even forgot to update it with the new Rocky 9 and SUSE 15.4 builds in https://github.com/rstudio/r-docker/pull/85. With GHA, we can take advantage of a few nice features:
   - The build matrix is now dynamic generated from the Makefile, so no more need to update the config when adding new platforms.
   - The build matrix is still parallelized by platform, but we could easily parallelize by R version as well if the build time gets longer, like in R-builds. Build times aren't an issue yet though.
   - The manual workflow trigger is more powerful in GHA than in CircleCI. The CircleCI console doesn't support parameters (last I checked), but GHA does, so we can now manually run builds with options to limit platforms/versions, disable publishing, and rebuild the full patch version matrix.
   - The CRON jobs to build R-devel daily or patch versions monthly are now configured in separate files (and can be manually triggered individually) for a little better organization. This is using GHA's [reusable workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows), which just requires an `on.workflow_call` section. Inputs have to be duplicated with `workflow_dispatch`, but you can refer to both inputs using `input.<param>` in the workflow.

2. Work around a Dockerfile caching bug with Docker BuildKit: https://github.com/moby/buildkit/issues/1368. There's a really bad bug where `docker build` can use the wrong Dockerfile for caching, even if the content is different, because the timestamp and file sizes of two Dockerfiles are the same. This caused https://github.com/rstudio/r-docker/issues/86 while I was testing publishing, which I thought would be safe enough.

The issue looks like this - note the incorrectly cached step of downloading R 3.5.3 in the `3.6-focal` build:
https://github.com/rstudio/r-docker/actions/runs/3041017846/jobs/4897712035
```sh
#6 writing image sha256:8810a636188861eb0f8641ebc05f760e9b467d962bbcadf83ff98c8864623d24 done
#6 naming to docker.io/rstudio/r-base:3.5-focal done
#6 DONE 3.0s
docker build -t rstudio/r-base:3.6-focal --build-arg BASE_IMAGE=rstudio/r-base 3.6/focal/.
#1 [internal] load build definition from Dockerfile
#1 transferring dockerfile: 32B done
#1 DONE 0.0s

#2 [internal] load .dockerignore
#2 transferring context: 2B done
#2 DONE 0.0s

#3 [internal] load metadata for docker.io/rstudio/r-base:focal
#3 DONE 0.0s

#4 [1/2] FROM docker.io/rstudio/r-base:focal
#4 DONE 0.0s

#5 [2/2] RUN wget https://cdn.rstudio.com/r/ubuntu-2004/pkgs/r-3.5.3_1_amd64.deb &&     apt-get update -qq &&     DEBIAN_FRONTEND=noninteractive apt-get install -f -y ./r-3.5.3_1_amd64.deb &&     ln -s /opt/R/3.5.3/bin/R /usr/bin/R &&     ln -s /opt/R/3.5.3/bin/Rscript /usr/bin/Rscript &&     ln -s /opt/R/3.5.3/lib/R /usr/lib/R &&     rm r-3.5.3_1_amd64.deb &&     rm -rf /var/lib/apt/lists/*
#5 CACHED

#6 exporting to image
#6 exporting layers done
#6 writing image sha256:8810a636188861eb0f8641ebc05f760e9b467d962bbcadf83ff98c8864623d24 done
#6 naming to docker.io/rstudio/r-base:3.6-focal done
```

This issue still isn't fixed as of the latest Docker version `20.10.17` installed by the GHA action, so I added one of the workarounds of piping the Dockerfile in through stdin. And added a test to ensure the R version of the image matches the tag to prevent this from happening again.

And finally, the builds do fail on openSUSE 15.4 right now because there's no R-devel build for SUSE 15.4. There's some issue with compilation, possibly a bug that may be fixed later, so I've left that as-is for now. As the R-devel binary and Docker image builds are retried daily, the builds may eventually pass if the issue is fixed. Otherwise, we could skip the build or look into a workaround for the R-devel problem.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        description: |
+          Include R patch versions?
+        required: false
+        default: false
+        type: boolean
+    secrets:
+      DOCKERHUB_USERNAME:
+        required: true
+      DOCKERHUB_TOKEN:
+        required: true
```
Comment (shepherdjerred) [inline .github/workflows/build.yml:55]: Why are these required if publishing can be skipped?

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+permissions:
+  contents: read
```
Comment (shepherdjerred) [inline .github/workflows/build.yml:58]: Isn't this permission granted by default?

File: .github/workflows/devel-daily.yml [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  schedule:
+    - cron: '0 6 * * *'
+  workflow_dispatch:
+
+permissions:
+  contents: read
+
+jobs:
+  docker-images:
+    uses: ./.github/workflows/build.yml
```
Comment (shepherdjerred) [inline .github/workflows/devel-daily.yml:13]: I didn't know this was possible, neat.

**Review by shepherdjerred — APPROVED** [review]
Looks good to me.

File: .github/workflows/build.yml (hunk 2 above)
Comment (glin) [reply .github/workflows/build.yml:58]: It is, but this restricts the permissions from the default of read/write for all scopes: https://github.blog/changelog/2021-04-20-github-actions-control-permissions-for-github_token/

I've noticed more workflows adding this explicitly and think it's a good idea. It's also possible to restrict permissions on the repo level, but is less obvious and probably doesn't work with forks.

File: .github/workflows/build.yml (hunk 1 above)
Comment (glin) [reply .github/workflows/build.yml:55]: There's a Docker login step that always gets run in the main workflow. In hindsight, I guess we could skip that step if publishing is skipped, but the secrets will be present regardless, and it is kind of useful to test that valid credentials are configured without having to publish an image.

Also wondering now, passing secrets may not even be necessary if the reusable workflow is in the same repo. I'll test if this config is even necessary and remove it if not.

File: .github/workflows/build.yml (hunk 2 above)
Comment (glin) [reply .github/workflows/build.yml:58]: It is, but this restricts the permissions from the default of read/write for all scopes: https://github.blog/changelog/2021-04-20-github-actions-control-permissions-for-github_token/

I've noticed more workflows adding this explicitly and think it's a good idea. It's also possible to restrict permissions on the repo level, but is less obvious and probably doesn't work with forks.

File: .github/workflows/build.yml (hunk 1 above)
Comment (glin) [reply .github/workflows/build.yml:55]: It's apparently required even for reusable actions in the same repo, so leaving this in.
